consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,394 @@
+# Specification for Celo IBFT protocol
+
+## Celo IBFT specification
+
+### High level overview
+
+The Celo IBFT protocol is a BFT (Byzantine Fault Tolerant) protocol that allows
+a group of participants to agree on the ordering of values by exchanging voting
+messages across a network. As long as less than 1/3rd of participants deviate
+from the protocol then the protocol should ensure that there is only one
+ordering of values that participants agree on and also that and that they can
+continue to agree on new values, i.e. they don't get stuck.
+
+Ensuring that only one ordering is agreed upon is referred to as 'safety' and
+ensuring that the participants can continue to agree on new values is referred
+to as 'liveness'.
+
+### System model
+
+We consider a group of `3f+1` participating computers (participants) that are
+able to communicate across a network. The group should be able to tolerate `f`
+failures without losing liveness or safety.
+
+All participants know the other participants in the system and it is assumed
+that all messages are signed in some way such that participants in the protocol
+know which participant a message came from and that messages cannot be forged.
+
+Only messages from participants are considered.
+
+For messages containing a value only messages with valid values are considered.
+
+Participants are able to broadcast messages to all other participants. In the
+case that a participant is off-line or somehow inaccessible they will not
+receive broadcast messages and there is no mechanism for these messages to be
+re-sent.
+
+We refer to a consensus instance to mean consensus for a specific height,
+consensus instances are independent (they have no shared state).
+
+#### Instance state
+`Sc - current participant state`\
+`Hc - current height`\
+`Rc - current round`\
+`Rd - desired round`\
+`Vc - currently proposed value`\
+`PCc - current prepared certificate`\
+
+### Algorithm
+See supporting [functions](#Functions) and [notation](#Appendix-1-Notation).
+```
+upon: FC_E
+  Hc ← Hc+1
+  Rc ← 0
+  Rd ← 0
+  Sc ← AcceptRequest
+  Vc ← nil
+  schedule onRoundChangeTimeout(Hc, 0) after roundChangeTimeout(0)
+
+upon: <R_E, Hc, V> && Sc = AcceptRequest
+  if Rc = 0 && isProposer(Hc, Rd) {
+    bc(<PP_T, Hc, 0, V, nil>)
+  }
+
+upon: <PP_T, Hc, Rd, V, RCC> from proposer(Hc, Rd) && Sc = AcceptRequest
```

Comment:
Giving names to each of these transitions would probably help to make it easier to understand